### PR TITLE
MLFlow Triton Plugin: Add support for s3 prefix and custom endpoint URL

### PR DESCRIPTION
**Pull request by yeahdongcn**
For the model repository, Triton server supports the following [S3 URI](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_repository.md#s3):

```bash
$ tritonserver --model-repository=s3://https://host:port/bucket/path/to/model/repository ...
```

But MLFlow Triton plugin currently only supports the bucket style like `s3://bucket` and there is no environment variable support to update the endpoint URL of S3 client. This PR adds support for S3 prefix and custom endpoint URL which aligns with the [server behavior](https://github.com/triton-inference-server/core/blob/main/src/filesystem.cc#L1464).

### Installation ✅
```bash
python setup.py install
```

#### Create deployment ✅

The `add_sub` model and `mlflow-meta.json` can be found in minio storage.

```bash
export TRITON_MODEL_REPO=s3://https://oss.vmware.com:443/release/cloud/models
mlflow deployments create -t triton --flavor triton --name add_sub -m models:/add_sub/1
```

#### Delete deployment ✅

```bash
mlflow deployments delete -t triton --flavor triton --name add_sub
```
### ONNX Flavor

```python
import mlflow.onnx
import onnx
model = onnx.load("examples/onnx_float32_int32_int32/1/model.onnx")
mlflow.onnx.log_model(model, "triton", registered_model_name="onnx_float32_int32_int32")

from mlflow.deployments import get_deploy_client
client = get_deploy_client('triton')
client.create_deployment('onnx_float32_int32_int32', 'models:/onnx_float32_int32_int32/1', flavor="onnx")
```

Files are correctly uploaded to s3.

```bash
❯ mc ls -r oss/release/cloud/models/onnx_float32_int32_int32
[2023-05-11 10:37:26 CST]   333B 1/model.onnx
[2023-05-11 10:37:27 CST]   208B mlflow-meta.json
```

### CODE REVIEW

**Comment by yeahdongcn** [comment]
@albertoperdomo2 @GuanLuo Could you please review this PR when you get a chance?

**Comment by albertoperdomo2** [comment]
Well, initially I would suggest to properly order the **Config** class, so the `__init__` method is the first one on the top. This make it so much easier to check which are class variables, which values are they initialized to, etc.

File: deploy/mlflow-triton-plugin/mlflow_triton/config.py [hunk 1]
```diff
@@ -24,17 +24,114 @@
...
             import boto3
-            self['s3'] = boto3.client('s3')  # boto3 handles AWS credentials
-            self['s3_bucket'] = self['triton_model_repo'].replace('s3://', '').replace('/', '')
+            # boto3 handles AWS credentials
+            self['s3'] = boto3.client(
+                's3', endpoint_url=endpoint_url)
+            self['s3_bucket'] = uri.bucket
+            self['s3_prefix'] = uri.prefix
+            self['triton_model_repo'] = 's3://{}'.format(
+                '/'.join(filter(None, [uri.bucket, uri.prefix])))
```
Comment (albertoperdomo2) [inline deploy/mlflow-triton-plugin/mlflow_triton/config.py:136]: What is the point of the `filter()` here?

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 2]
```diff
@@ -189,8 +190,9 @@ def list_deployments(self):
                 if 's3' in self.server_config:
                     meta_dict = ast.literal_eval(self.server_config['s3'].get_object(
                         Bucket=self.server_config['s3_bucket'],
-                        Key=d['name'] + f'/{_MLFLOW_META_FILENAME}',
-                        )['Body'].read().decode('utf-8'))
+                        Key='/'.join(filter(
```
Comment (albertoperdomo2) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:193]: Same here: `filter(None, <list>)` is basically returning `<list>`, so why use if in the first place?

**Review by albertoperdomo2 — CHANGES_REQUESTED** [review]
Another thing is that, we all use different linters/formatters for our code, so I would advise to not push any changes related to formatting tabs and so on, because it make it difficult to review your actual changes. But this is an opinion, @GuanLuo might think something else.

**Comment by yeahdongcn** [comment]
> Another thing is that, we all use different linters/formatters for our code, so I would advise to not push any changes related to formatting tabs and so on, because it make it difficult to review your actual changes. But this is an opinion, @GuanLuo might think something else.

I was not intended to reformat (vscode plugin did), I can revert these changes.

File: deploy/mlflow-triton-plugin/mlflow_triton/config.py (hunk 1 above)
Comment (yeahdongcn) [reply deploy/mlflow-triton-plugin/mlflow_triton/config.py]: `os.path.join` works just fine. Please see the latest commit, thanks.

**Comment by yeahdongcn** [comment]
> Well, initially I would suggest to properly order the **Config** class, so the `__init__` method is the first one on the top. This make it so much easier to check which are class variables, which values are they initialized to, etc.

This is also updated in the latest commit. Thanks.

**Review by albertoperdomo2 — COMMENTED** [review]
LGTM! 🚀

**Comment by albertoperdomo2** [comment]
I'll run the CI locally later to see if all the tests pass. We should setup a CI pipeline in the near future though.

**Comment by yeahdongcn** [comment]
> I'll run the CI locally later to see if all the tests pass. We should setup a CI pipeline in the near future though.

That will be great. If you have the ONNX flavor test environment, please also feel free to update this PR. Thanks.

**Review by GuanLuo — APPROVED** [review]
@yeahdongcn thanks for your contribution, and @albertoperdomo2 thank you for the review. I will kick off the CI for sanity testing and then the change should be good to be merged.

**Comment by GuanLuo** [comment]
Regarding the code formatting, we do have contribution guideline [here](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md#coding-convention), you can find the script used [here](https://github.com/triton-inference-server/common/tree/main/tools) which also provides a git pre-commit hook that you may use to trigger the formatter

**Comment by albertoperdomo2** [comment]
Probably not the right place to discuss it but I would love to chat with you @GuanLuo to see what would be needed to set up a basic CI pipeline so the CI/CD process would be faster. Hasn't it be configured due to technical limitations or time constraints? Feel free to hit me up by email or Twitter.

**Comment by GuanLuo** [comment]
@yeahdongcn The [L0_mlflow](https://github.com/triton-inference-server/server/tree/main/qa/L0_mlflow) failed on using ONNX flavor with S3, I think your change will need to be applied for ONNX flavor as well to pass the test. I can look into it when I get to it but it will probably be near the end of the week. If you don't mind, you can also test it yourself following the instruction [here](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/test.md) to prepare a test-ready docker image.

**Comment by GuanLuo** [comment]
@albertoperdomo2 We do want to figure out a public-facing CI integration but that is a long shot, mostly due to our current CI is tightly coupled with internal infrastructure and thus need to un-mangle it first. @mc-nv may have more insight on the matter.

As of now, the guideline for running the tests is to build the QA container locally as mentioned above.

**Comment by yeahdongcn** [comment]
> @yeahdongcn The [L0_mlflow](https://github.com/triton-inference-server/server/tree/main/qa/L0_mlflow) failed on using ONNX flavor with S3, I think your change will need to be applied for ONNX flavor as well to pass the test. I can look into it when I get to it but it will probably be near the end of the week. If you don't mind, you can also test it yourself following the instruction [here](https://github.com/triton-inference-server/server/blob/main/docs/customization_guide/test.md) to prepare a test-ready docker image.

Sure. Will look into the test code.

**Comment by yeahdongcn** [comment]
@GuanLuo Please see the latest commit, thanks.

BTW, I'm unable to run `server/qa/L0_mlflow/test.sh` directly. So I followed the `server/deploy/mlflow-triton-plugin/README.md` to load an ONNX model and then deploy it to triton server.

```python
import mlflow.onnx
import onnx
model = onnx.load("examples/onnx_float32_int32_int32/1/model.onnx")
mlflow.onnx.log_model(model, "triton", registered_model_name="onnx_float32_int32_int32")

from mlflow.deployments import get_deploy_client
client = get_deploy_client('triton')
client.create_deployment('onnx_float32_int32_int32', 'models:/onnx_float32_int32_int32/1', flavor="onnx")
```

The above code should be similar to `plugin_test.py` and I also checked the s3, the files are correctly uploaded:

```bash
❯ mc ls -r oss/release-ci/cloud/models/onnx_float32_int32_int32
[2023-05-11 10:37:26 CST]   333B 1/model.onnx
[2023-05-11 10:37:27 CST]   208B mlflow-meta.json
```

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 3]
```diff
@@ -359,64 +363,68 @@ def _get_copy_paths(self, artifact_path, name, flavor):
...
+
                         if flavor == "onnx":
                             s3_path = os.path.join(
-                                    copy_paths[key]['to'].replace(
-                                    self.server_config['triton_model_repo'], ''), 
-                                    filename,
-                                    ).replace('/', '', 1)
+                                self.server_config['s3_prefix'],
+                                copy_paths[key]['to'].replace(
+                                    self.server_config['triton_model_repo'], '').replace('/', '', 1),
```
Comment (GuanLuo) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:393]:
```suggestion
                                    self.server_config['triton_model_repo'], ''),
```
Seems like the replace here will strip out the slash for version directory, i.e.
`onnx_model_with_files/1/model.onnx -> onnx_model_with_files1/model.onnx`

L0_mflow passed after removing it, @albertoperdomo2 I wonder if you remember why the replace was needed?

Comment (yeahdongcn) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:393]: Thanks for the review and test @GuanLuo 
Just did some experiments to figure out why replace is used previously.

```python
export TRITON_MODEL_REPO=s3://https://oss.vmware.com:443/release/cloud/models
>>> client.create_deployment('onnx_float32_int32_int32', 'models:/onnx_float32_int32_int32/1', flavor="onnx")
print(copy_paths[key]['to'])
s3://release/cloud/models/onnx_float32_int32_int32/1

print(self.server_config['triton_model_repo'])
s3://release/cloud/models

print(copy_paths[key]['to'].replace(self.server_config['triton_model_repo'], ''))
/onnx_float32_int32_int32/1 ==> `replace` is used in this case to trim the first occupation of `/`

print(filename)
model.onnx

print(s3_path)
cloud/models/onnx_float32_int32_int32/1/model.onnx
```

```python
export TRITON_MODEL_REPO=s3://bucket_name
>>> client.create_deployment('onnx_float32_int32_int32', 'models:/onnx_float32_int32_int32/1', flavor="onnx")
print(copy_paths[key]['to'])
s3://bucket_name/onnx_float32_int32_int32/1

print(self.server_config['triton_model_repo'])
s3://bucket_name/

print(copy_paths[key]['to'].replace(self.server_config['triton_model_repo'], ''))
onnx_float32_int32_int32/1

print(filename)
model.onnx

print(s3_path)
onnx_float32_int32_int32/1/model.onnx
```

**Review by albertoperdomo2 — CHANGES_REQUESTED** [review]
It looks good to me, but you need to make sure the tests pass. Other than the mentioned comments, as soon as the tests pass, it's ready to merge for me. You should probably add a few URL/test cases [in here](https://github.com/triton-inference-server/server/blob/main/qa/L0_mlflow/test.sh#L199), and make sure it passes the old ones (with the standard/simple URL) and the new ones you add.

File: deploy/mlflow-triton-plugin/mlflow_triton/config.py [hunk 4]
```diff
@@ -24,17 +24,114 @@
...
+import re
+from collections import namedtuple
+from mlflow.exceptions import MlflowException
 
 
 class Config(dict):
 
     def __init__(self):
         super().__init__()
+        self.s3_regex = re.compile(
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply deploy/mlflow-triton-plugin/mlflow_triton/config.py:36]: Thanks for the suggestion. It has been resolved in the latest commit.

**Comment by yeahdongcn** [comment]
> It looks good to me, but you need to make sure the tests pass. Other than the mentioned comments, as soon as the tests pass, it's ready to merge for me. You should probably add a few URL/test cases [in here](https://github.com/triton-inference-server/server/blob/main/qa/L0_mlflow/test.sh#L199), and make sure it passes the old ones (with the standard/simple URL) and the new ones you add, because the current bucket URL used for the tests is just `s3://<bucket-name>`.

No problem. I will add some tests in a new PR.

**Review by GuanLuo — APPROVED** [review]
@yeahdongcn Just verified the commit passed the test, let me know if the PR is ready to be merged on you end

File: deploy/mlflow-triton-plugin/mlflow_triton/config.py [hunk 5]
```diff
@@ -34,7 +37,102 @@
...
+            raise MlflowException("Invalid bucket name: '" + path + "'")
+
+        # Remove leading slashes
+        ltrim_length = len(path) - len(path.lstrip('/'))
+        if ltrim_length == len(path):
+            raise MlflowException("Invalid bucket name: '" + path + "'")
+
+        # Remove extra internal slashes
+        true_path = path[ltrim_length:rtrim_length + 1]
+        slash_locations = []
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply deploy/mlflow-triton-plugin/mlflow_triton/config.py:127]: Fixed.

**Comment by yeahdongcn** [comment]
> @yeahdongcn Just verified the commit passed the test, let me know if the PR is ready to be merged on you end

Thanks for the verification @GuanLuo 

Just resolved a `unused local variable issue found by [github-code-scanning](https://github.com/apps/github-code-scanning), the code change is just removing `slash_locations`.

So please merge it. Thank you~

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**